server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -3433,12 +3433,23 @@ public void activateWithPrimaryContext(final ReplicationTracker.PrimaryContext p
             + "] does not contain relocation target ["
             + routingEntry()
             + "]";
-        assert getLocalCheckpoint() == primaryContext.getCheckpointStates().get(routingEntry().allocationId().getId()).getLocalCheckpoint()
-            || indexSettings().getTranslogDurability() == Durability.ASYNC : "local checkpoint ["
-                + getLocalCheckpoint()
-                + "] does not match checkpoint from primary context ["
-                + primaryContext
-                + "]";
+        if (isRemoteStoreEnabled()) {
+            assert getLocalCheckpoint() == primaryContext.getCheckpointStates()
```

Comment:
I'm having a hard time parsing this and while you did an admirable job explaining it in the PR description I'm not wrapping my head around what I read hear matching it.  I'd recommend rewriting this into different bool / locals that help document what is being checked from where and why.

Reply:
Refactored. Please check.